src/__tests__/acceptance/user.controller.acceptance.ts

```diff
@@ -0,0 +1,159 @@
+        .expect(401);
+      expect(response.status).to.equal(401);
+    });
+
+    it('Fails when DELETE /users/{id} without authorization', async () => {
+      const response = await client
+        .get('/users/{id}')
+        .set('Authorization', 'noBearer' + 'not.thing.here')
+        .expect(401);
+      expect(response.status).to.equal(401);
+    });
+  });
+
+  describe('Successful Authorizations', () => {
+    it('Success on invoking GET /users/{id}', async () => {
+      userInfo = await createUser({
+        email: 'obed@gmail.com',
+        name: 'Obed',
+        dob: '2020-08-08T12:11:31.076Z',
+      });
+
+      let response = await client.post('/auth/login').send({email: userInfo.email, password: password}).expect(200);
+      token = response.body.token;
+
+      response = await client
+        .get(`/users/${userInfo.id}`)
+        .set('Authorization', 'Bearer ' + token)
+        .expect(200);
+
+      expect(response.body.id).to.equal(userInfo.id);
+    });
+
+    it('Success on invoking GET /users', async () => {
+      userInfo = await createUser({
+        email: 'obed@gmail.com',
+        name: 'Obed',
+        address: 'sh',
+        dob: '2020-08-08T12:11:31.076Z',
+      });
+
+      let response = await client.post('/auth/login').send({email: userInfo.email, password: password}).expect(200);
+      token = response.body.token;
+
+      response = await client
+        .get(`/users`)
+        .send({address: 'Shanghai'})
+        .set('Authorization', 'Bearer ' + token)
+        .expect(200);
+      expect(response.status).to.equal(200);
+      expect(response.body).to.not.be.empty();
+    });
+
+    it('Success on invoking PATCH /users/{id}', async () => {
+      userInfo = await createUser({
+        email: 'obed@gmail.com',
+        name: 'Obed',
+        address: 'sh',
+        dob: '2020-08-08T12:11:31.076Z',
+      });
+
+      let response = await client.post('/auth/login').send({email: userInfo.email, password: password}).expect(200);
+      token = response.body.token;
+
+      response = await client
+        .patch(`/users/${userInfo.id}`)
+        .send({address: 'Shanghai'})
+        .set('Authorization', 'Bearer ' + token)
+        .expect(204);
+      expect(response.status).to.equal(204);
+    });
+
+    it('Success on DELETE /users/{id}', async () => {
+      userInfo = await createUser({
+        email: 'obed@gmail.com',
+        name: 'Obed',
+        dob: '2020-08-08T12:11:31.076Z',
+      });
+      let response = await client.post('/auth/login').send({email: userInfo.email, password: password}).expect(200);
+      token = response.body.token;
+
+      response = await client
+        .delete(`/users/${userInfo.id}`)
+        .set('Authorization', 'Bearer ' + token)
+        .expect(204);
+      expect(response.status).to.equal(204);
+    });
+  });
+
+  async function cleanDB() {
+    await userRepo.deleteAll();
+  }
+
+  async function createUser(data: object) {
```

Comment:
In acceptance tests of the application from the end users' perspective, only application's web APIs can be used to test, the end users cannot call application's internal methods.

So in the acceptance tests of the controller, method of the controller layer is preferred over the internal method of userRepo.

Reply:
Yes, you are right.